automation/TaskManager.py

```diff
@@ -498,11 +518,18 @@ def _issue_command(self, browser: Browser,
                 browser.restart_required = True
                 self.logger.debug("BROWSER %i: Browser restart required" % (
                     browser.crawl_id))
+
             else:
                 with self.threadlock:
                     self.failurecount = 0
 
             if browser.restart_required:
+                self.sock.send((RECORD_TYPE_SPECIAL, {
```

Comment:
Note that `browser.restart_required` is also set to True by the memory monitor, which can lead to bugs and might be the cause of the duplicate records you've noticed. See:
https://github.com/mozilla/OpenWPM/blob/9c8e2ea8678da2d0916bb7d9a4855341b20a9bae/automation/TaskManager.py#L209-L232

The browser may successfully complete the last command, but also run over the memory limit at the same time. The extension will send back success and then this code will send back failure (which is incorrect, the visit *did* finish). We can mitigate this by having the memory monitor use a different browser property (e.g., `browser.memory_exceeded`) and perhaps `restart_required` can then just be a local variable.

Reply:
Also is finalize command always sent twice, or only sometimes?

Reply:
The finalize command is only very rarely called twice. And it only happened after the crawl had been running for a while. I should investigate if it relates to the memory watchdog

Reply:
Also if we have a false positive (as in mark a visit incomplete when it's not) it isn't a big deal as the job will just retry and run to completion